compat/mingw.c

```diff
@@ -3029,6 +3030,60 @@ int readlink(const char *path, char *buf, size_t bufsiz)
 	return min(bufsiz, len);
 }
 
+static int wreadlink_len(const WCHAR *wpath, DWORD *ptag)
+{
+	HANDLE handle;
+	WCHAR *wbuf;
+	REPARSE_DATA_BUFFER *b = alloca(MAXIMUM_REPARSE_DATA_BUFFER_SIZE);
+	DWORD dummy;
+	int len;
+
+	/* read reparse point data */
+	handle = CreateFileW(wpath, 0,
```

Comment:
In our performance tests, `CreateFile()` was _substantially_ slower than `FindFirstFile()`. This might not sound like a big thing, but when you have tens of thousands of files (or even [several million files](https://devblogs.microsoft.com/bharry/the-largest-git-repo-on-the-planet/)), it does matter.

Noticing that symbolic links are relatively rare, would it not make more sense to replace the `MAX_LONG_PATH` constant by a function that determines the length of a symbolic link (and is called _only_ for symbolic links)?

Reply:
And while we're already at that point where we want to read the length of the symlink, would it not make more sense to use the [`readlink()`](https://github.com/git-for-windows/git/blob/v2.27.0-rc2.windows.1/compat/mingw.c#L2974-L3033) function rather than duplicating (part of) its logic?

Reply:
Thank you for the quick review.

> In our performance tests, `CreateFile()` was substantially slower than `FindFirstFile()`.

The `FindFirstFileW`, `FindClose` sequence makes the following calls:

- `NtOpenFile` / `IoCreateFile` on the **parent directory** of the supplied file path.
- `NtQueryDirectoryFile` with the requested file name or mask of the supplied file path.
- `NtClose`

The `CreateFileW`, `DeviceIoControl`, `CloseHandle` sequence makes the following calls:

- `NtCreateFile` / `IoCreateFile` on the supplied file path.
- `NtFsControlFile` with `FSCTL_GET_REPARSE_POINT`.
- `NtClose`.

So if you have observed performance differences it is perhaps because when accessing multiple files in the same directory the file system has a better chance to cache directory contents in the `FindFirstFileW` sequence.

Even more importantly for our case, `FindFirstFileW` does not give us enough information to fill in the symlink `stat`. It gives us the reparse tag, but does not give us the length of the symlink. So querying the reparse point for its data and computing the length in that way becomes unavoidable.

> Noticing that symbolic links are relatively rare, would it not make more sense to replace the `MAX_LONG_PATH` constant by a function that determines the length of a symbolic link (and is called only for symbolic links)?

I agree. But please note that this happens already: the expensive reparse point computation happens only when the `if (fdata.dwFileAttributes & FILE_ATTRIBUTE_REPARSE_POINT)` condition is true.

> And while we're already at that point where we want to read the length of the symlink, would it not make more sense to use the `readlink()` function rather than duplicating (part of) its logic?

I agree and that is how I implemented it at first. Unfortunately `readlink` does not return the actual reparse tag that is used in the computation of `st_mode` by `file_attr_to_st_mode`. Without that special `wreadlink_len` function we would have to issues both a `FindFirstFileW` and a `readlink` to properly `stat` a symlink.

Reply:
BTW, one improvement in the submitted commit is the following:

Currently the `wreadlink_len` returns -1 it if encounters a reparse point that is not a symbolic link or a mount point; this causes `mingw_lstat` to also return -1. Perhaps such cases should be handled more gracefully (e.g. by returning the prior size of 4096 for compatibility purposes).

Reply:
> Without that special `wreadlink_len` function we would have to issue both a `FindFirstFileW` and a `readlink` to properly `stat` a symlink.

Well, in the current state, there is a lot of duplicated code.

The common strategy in Git's commit history is to refactor out a helper function that can be used from multiple call sites. In this case, it would be factoring out that part of `readlink()` that determines e.g. `wbuf` and `len`, then calling that from `readlink()` _and_ from `mingw_lstat()`.

That way, code is not duplicated, and as a consequence there is not even a possibility for one of the two copies of the code to become stale, as there is only a single copy of the code to begin with.

Reply:
Understood. I will look into making the necessary changes and resubmit.

Reply:
Thank you!